### ISSUE TITLE
Jaw Open morph doesn't open the jaw anymore after the roundtrip.

### ISSUE THREAD

**Issue by loicbramoulle** [soupday/cc_blender_tools#124]
Hi, I'm facing this JawOpen shenanigans, where the shape/morph does open the jaw at start in CC4 & iClone, but once the character get sent to Blender, it only deforms, waiting for the JawBone to open the jaw. Maybe in CC4 & iClone the JawOpen morph does drive the JawBone, BUT once the characters gets back in CC4 with the pipeline/autoSetup, the JawOpen morph stops opening the jaw/driving the JawBone.
Maybe I'm missing something, but this seems to be a tiny showstopper that prevents the whole pipeline to be viable.
Thanks for any help.

**Comment by loicbramoulle** [comment]
It is actually the same thing for the eyes meshes, once the characters gets back in CC4, moving the eyes moves only the eyelids, the eye globes stays fixed, as I guess the link between the facial shape and the bone is broken by the pipeline.
In Blender the eyeUp/downLleft/right shapes don't even exist on the eyes, so it's indeed because they're never moving with a shape, but with the eye bone rotation. Only with ARKit the eye globes, teeth & jaw are working fully with shapes, no bones.

So is there a way to repair this broken link between the facial shapes and the bones once the character gets back in CC4 ?

**Comment by loicbramoulle** [comment]
Ok it's indeed this, the facial profile is broken during the roundtrip, so the facial shapes stops driving the bones. It's possible to re-create these links by hand, but it's pretty long annoying and buggy it seems, my JawBone skinning is all good in Blender, but totally broken once re-imported via the autoSetup, so I can't recreate the links by hand:

![image](https://user-images.githubusercontent.com/49880198/214215869-0b9c2ae6-1cdb-423b-bdd5-f93807dd37ec.png)

**Comment by soupday** [comment]
Looks like it's a bug in Character Creator in the batch import of blend shapes. Any blend shapes with associated bone movements (i.e. The Jaw and EyeLook categories) loose the connection with the bone. It does this even doing it by hand in the Facial Profile Editor. So it looks like it's not possible to update those blend shapes without breaking the bone movements.

So, for now, all I can do is exclude those categories (Jaw and EyeLook) from the batch import.

**Comment by soupday** [comment]
Blender side, I can put in drivers to either drive the jaw bone rotation from the Jaw_Open blend shape, or the blend shape from the jaw bone rotation.

Of course that gets very complicated very quickly once you try to rigify it...

**Comment by loicbramoulle** [comment]
Damn ok, it's odd that I did not have this issue last year when doing the same thing, redoing the whole chcracter in blender and animating it in iClone, eyes & jaw just worked, driven by bones once I exported animation data from iClone to render in Blender.
I've just seen this message on a video, I guess that's this link breaking issue it refers to ?
![image](https://user-images.githubusercontent.com/49880198/214340461-a7f43917-0205-420b-bb39-ce47004d1401.png)
Anyway, it seems pretty important to find a workaround as it invalidates totally the whole process of making a Reallusion character your own, if eyes & jaw are broken on import. Maybe for now I can try to create the link by hand like above, and cleanup manually the skinning ? The issue is that I would reimport and update the character multiple times every time I see thing to improve while animating in iClone, so redoing all eyes & jaw drivers by hand kinda prevents that.
Thanks a lot for your answer ! I will look if I find any workaround on my side.

**Comment by loicbramoulle** [comment]
yes, on the Huntress short film the facial links didn't break last year in May. 
![image](https://user-images.githubusercontent.com/49880198/214710535-29ea5763-ee15-4015-b5b8-2657ef1731f2.png)

Now when I reimport the fbx from last year CC4 crashes, so maybe something changed in the plugin ?

**Comment by loicbramoulle** [comment]
hum cc4 crashes even trying to import a new fbx exported with the pipeline, the check is all green/pass, but cc4 shuts down afterward.

**Comment by loicbramoulle** [comment]
So.. it's not possible to paint skin weight on a character's body in CC ?.. how do you .. I mean.. how.. what.
Like if the pipeline break the jaw weights, and the link, I can recreate those by hand each time I update the character in blender, but if CC4 doesn't have anything to simply edit the skinning of a character (only props..) then.. we can't use this pipeline. Apart for characters without a face...

**Comment by loicbramoulle** [comment]
Ok so maybe we can all just send Kevin to iC8, animate him, and then only in blender we just remap on our version where it's an actual character..

**Comment by soupday** [comment]
> So.. it's not possible to paint skin weight on a character's body in CC ?.. how do you .. I mean.. how.. what. Like if the pipeline break the jaw weights, and the link, I can recreate those by hand each time I update the character in blender, but if CC4 doesn't have anything to simply edit the skinning of a character (only props..) then.. we can't use this pipeline. Apart for characters without a face...

It doesn't break the weights, the animation link between the expression shape and bone movements is broken in Character Creator when importing those expressions. It's a bug in Character Creator and a serious one, because it breaks most of the facial animations. There's no implicit link defined anywhere, Character Creator does this internally so it breaks it internally as well.

**Comment by soupday** [comment]
I've updated the [plugin](https://github.com/soupday/CC4-Blender-Tools-Plugin/releases/tag/1.1.3)

For now, it doesn't update those facial expressions when re-importing the character so it shouldn't break those animations. But this does mean that you can't change the facial expressions in the Jaw, Head and EyeLook categories.

**Comment by soupday** [comment]
> hum cc4 crashes even trying to import a new fbx exported with the pipeline, the check is all green/pass, but cc4 shuts down afterward.

The FBX import function tends to crash after multiple imports. I find it's most stable after closing and relaunching. Also close everything else, the import process uses a **whole lot** of video memory and can crash when it runs out.

**Comment by loicbramoulle** [comment]
Wow thanks for all these info Soupday ! I see more clearly what's going on, and how this pipeline is developped.
I think that should be all good is the jaw+head+eye look is not imported and stays just as it is by default, it's really more the smile, blink, pucker, that deserve to be sculpted by hand in blender to follow the character design.

I did have a skin weight issue as you see in the first screenshot here, I think it's only the jaw weight that broke thought, but yes on cc3+ characters it's forbiden to edit the skin weight apparently, arbitrarily, and humanoid type characters it's possible, but converting a cc3+ to humanoid destroys all facial shapes it seems, or they're still there, but they don't react to animation anymore .. so I fixed the skinning by hand, but then use the previous version and animated with the destroyed skinning for now.

Thanks again.

**Comment by loicbramoulle** [comment]
Just tested the new version, it worked on a default Kevin, but on a more customized version there's still the skin weight bug on the jaw/lip, and the eye bones that seems off-centered at the back of the head..
![image](https://user-images.githubusercontent.com/49880198/215031161-462059f1-a062-4ec0-915e-ad5531e5e528.png)

**Comment by loicbramoulle** [comment]
Ah thanks, yes I saw that video last week, and the hannibal lecter one hehe, and that makes a lot of sense also for more animation once in Blender, to do everything in the graph editor, instead of alternating between pose mode and shape key edition.
I nearly installed the previous version of the pluggin again, but for now I think it's simpler to leave jaw/teeth/eyes shapes as they are, and keep the armature animation for jaw and eyes. it's a bit of a pain for animation in blender though.
Also, the "open mouth as morph" in iClone file>export to fbx, doesn't work, it just doesn't seem to bake the jaw bone on the shape, the bone is still animated. but anycase it's not ideal to bake that

**Comment by ikrowni** [comment]
Is a fix for this issue available or are we still waiting for reallusion?

**Comment by soupday** [comment]
The only current fix is not to update those facial expressions that have corresponding bone movements.

You can either:

- In Blender delete any shapekey on the body that starts with Jaw_ , EyeLook_ or Head_ .
- Or use the current version of the CC4 Plugin (1.1.3) which blocks these blend shapes from updating.
- If none of that works you can "Reset Target Morph" on the facial expressions in the CC4 Facial Profile editor. (from the little gear icon in the top right of each expression. Everything in the Jaw, Head and EyeLook categories will need to be reset.

**Comment by loicbramoulle** [comment]
Aaah ok if we delete these shapes on the mesh in blender it will just assume it needs to use the default ones in CC4 but just for these, jaw & eyes (& head yes)
And nice one for the third point, resetting these shapes to the default one by one with the gear icon, I think that's plenty enough solutions.
